File: algos/unionfind_py/unionfind_test.py
import unittest
import sys
import os
import importlib.util
from typing import Protocol


class UnionFindProtocol(Protocol):
    def find(self, x: int) -> int: ...
    def union(self, x: int, y: int) -> None: ...
    def connected(self, x: int, y: int) -> bool: ...
    def count(self) -> int: ...


def load_unionfind_module():
    """Dynamically load the most recent unionfind_*.py file"""
    parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    files = [f for f in os.listdir(parent_dir) if f.startswith('unionfind_') and f.endswith('.py')]

    if not files:
        raise FileNotFoundError("No unionfind_*.py file found")

    # Get the most recent file
    latest_file = sorted(files)[-1]
    module_path = os.path.join(parent_dir, latest_file)

    spec = importlib.util.spec_from_file_location("unionfind_impl", module_path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    return module.UnionFind


class TestUnionFind(unittest.TestCase):

    def setUp(self):
        self.UnionFind = load_unionfind_module()

    def test_single_element(self):
        uf = self.UnionFind(1)
        self.assertEqual(uf.find(0), 0)
        self.assertEqual(uf.count(), 1)

    def test_two_separate_elements(self):
        uf = self.UnionFind(2)
        self.assertFalse(uf.connected(0, 1))
        self.assertEqual(uf.count(), 2)

    def test_basic_union(self):
        uf = self.UnionFind(3)
        uf.union(0, 1)
        self.assertTrue(uf.connected(0, 1))
        self.assertFalse(uf.connected(1, 2))
        self.assertEqual(uf.count(), 2)

    def test_chain_unions(self):
        uf = self.UnionFind(4)
        uf.union(0, 1)
        uf.union(1, 2)
        self.assertTrue(uf.connected(0, 2))
        self.assertFalse(uf.connected(0, 3))
        self.assertEqual(uf.count(), 2)

    def test_complex_operations(self):
        uf = self.UnionFind(6)
        uf.union(0, 1)
        uf.union(2, 3)
        uf.union(4, 5)
        self.assertEqual(uf.count(), 3)

        uf.union(1, 3)
        self.assertTrue(uf.connected(0, 2))
        self.assertFalse(uf.connected(0, 4))
        self.assertEqual(uf.count(), 2)

    def test_path_compression(self):
        uf = self.UnionFind(10)

        # Create a long chain: 0-1-2-3-4-5-6-7-8-9
        for i in range(9):
            uf.union(i, i + 1)

        # All should be connected
        for i in range(10):
            for j in range(i + 1, 10):
                self.assertTrue(uf.connected(i, j),
                              f"Expected {i} and {j} to be connected")

        # Should have only 1 component
        self.assertEqual(uf.count(), 1)

    def test_disjoint_sets(self):
        uf = self.UnionFind(8)

        # Create two disjoint sets: {0,1,2,3} and {4,5,6,7}
        uf.union(0, 1)
        uf.union(1, 2)
        uf.union(2, 3)

        uf.union(4, 5)
        uf.union(5, 6)
        uf.union(6, 7)

        # Test within-set connections
        test_connections = [
            (0, 1, True), (0, 2, True), (0, 3, True),
            (1, 2, True), (1, 3, True), (2, 3, True),
            (4, 5, True), (4, 6, True), (4, 7, True),
            (5, 6, True), (5, 7, True), (6, 7, True),
            (0, 4, False), (1, 5, False), (2, 6, False), (3, 7, False),
        ]

        for x, y, expected in test_connections:
            result = uf.connected(x, y)
            self.assertEqual(result, expected,
                           f"Connected({x}, {y}) = {result}, want {expected}")

        self.assertEqual(uf.count(), 2)

    def test_sequential_operations(self):
        uf = self.UnionFind(5)

        # Initial state - all separate
        self.assertEqual(uf.count(), 5, "Initial count should be 5")

        # Check not connected
        self.assertFalse(uf.connected(0, 1), "0 and 1 should not be connected initially")

        # Union first pair
        uf.union(0, 1)

        # Check now connected
        self.assertTrue(uf.connected(0, 1), "0 and 1 should be connected after union")

        # Count should decrease
        self.assertEqual(uf.count(), 4, "Count should be 4 after first union")

        # Check Find returns same root for connected elements
        root0 = uf.find(0)
        root1 = uf.find(1)
        self.assertEqual(root0, root1, "Find(0) and Find(1) should return same root")

        # Union another pair
        uf.union(2, 3)

        # Count should decrease again
        self.assertEqual(uf.count(), 3, "Count should be 3 after second union")

        # Elements 0,1 should not be connected to 2,3
        self.assertFalse(uf.connected(0, 2), "0 and 2 should not be connected")

    def test_idempotent(self):
        uf = self.UnionFind(3)

        # Union the same pair multiple times
        uf.union(0, 1)
        count1 = uf.count()

        uf.union(0, 1)  # Should have no effect
        count2 = uf.count()

        uf.union(1, 0)  # Reverse order, should have no effect
        count3 = uf.count()

        self.assertEqual(count1, count2, "Count should remain same after repeated union")
        self.assertEqual(count2, count3, "Count should remain same after reversed union")
        self.assertEqual(count1, 2, "Count should be 2")


if __name__ == '__main__':
    unittest.main()

File: algos/archived/unionfind/unionfind_test.go
package unionfind

import "testing"

type unionFindInterface interface {
	Find(x int) int
	Union(x, y int)
	Connected(x, y int) bool
	Count() int
}

var basicTestCases = []struct {
	name       string
	size       int
	operations []operation
	expected   expectedResults
}{
	{
		name: "single element",
		size: 1,
		operations: []operation{
			{op: "find", x: 0, expected: 0},
			{op: "count", expected: 1},
		},
	},
	{
		name: "two separate elements",
		size: 2,
		operations: []operation{
			{op: "connected", x: 0, y: 1, expected: false},
			{op: "count", expected: 2},
		},
	},
	{
		name: "basic union",
		size: 3,
		operations: []operation{
			{op: "union", x: 0, y: 1},
			{op: "connected", x: 0, y: 1, expected: true},
			{op: "connected", x: 1, y: 2, expected: false},
			{op: "count", expected: 2},
		},
	},
	{
		name: "chain unions",
		size: 4,
		operations: []operation{
			{op: "union", x: 0, y: 1},
			{op: "union", x: 1, y: 2},
			{op: "connected", x: 0, y: 2, expected: true},
			{op: "connected", x: 0, y: 3, expected: false},
			{op: "count", expected: 2},
		},
	},
	{
		name: "complex operations",
		size: 6,
		operations: []operation{
			{op: "union", x: 0, y: 1},
			{op: "union", x: 2, y: 3},
			{op: "union", x: 4, y: 5},
			{op: "count", expected: 3},
			{op: "union", x: 1, y: 3},
			{op: "connected", x: 0, y: 2, expected: true},
			{op: "connected", x: 0, y: 4, expected: false},
			{op: "count", expected: 2},
		},
	},
}

type operation struct {
	op       string
	x, y     int
	expected interface{}
}

type expectedResults struct {
	finalCount int
	groups     [][]int
}

func testUnionFindBasic(t *testing.T, newUF func(n int) unionFindInterface) {
	for _, tc := range basicTestCases {
		t.Run(tc.name, func(t *testing.T) {
			uf := newUF(tc.size)

			for i, op := range tc.operations {
				switch op.op {
				case "find":
					result := uf.Find(op.x)
					if expected, ok := op.expected.(int); ok && result != expected {
						t.Errorf("operation %d: Find(%d) = %d, want %d", i, op.x, result, expected)
					}
				case "union":
					uf.Union(op.x, op.y)
				case "connected":
					result := uf.Connected(op.x, op.y)
					if expected, ok := op.expected.(bool); ok && result != expected {
						t.Errorf("operation %d: Connected(%d, %d) = %v, want %v", i, op.x, op.y, result, expected)
					}
				case "count":
					result := uf.Count()
					if expected, ok := op.expected.(int); ok && result != expected {
						t.Errorf("operation %d: Count() = %d, want %d", i, result, expected)
					}
				}
			}
		})
	}
}

func testUnionFindPathCompression(t *testing.T, newUF func(n int) unionFindInterface) {
	uf := newUF(10)

	// Create a long chain: 0-1-2-3-4-5-6-7-8-9
	for i := 0; i < 9; i++ {
		uf.Union(i, i+1)
	}

	// All should be connected
	for i := 0; i < 10; i++ {
		for j := i + 1; j < 10; j++ {
			if !uf.Connected(i, j) {
				t.Errorf("Expected %d and %d to be connected", i, j)
			}
		}
	}

	// Should have only 1 component
	if count := uf.Count(); count != 1 {
		t.Errorf("Count() = %d, want 1", count)
	}
}

func testUnionFindDisjointSets(t *testing.T, newUF func(n int) unionFindInterface) {
	uf := newUF(8)

	// Create two disjoint sets: {0,1,2,3} and {4,5,6,7}
	uf.Union(0, 1)
	uf.Union(1, 2)
	uf.Union(2, 3)

	uf.Union(4, 5)
	uf.Union(5, 6)
	uf.Union(6, 7)

	// Test within-set connections
	testConnections := []struct {
		x, y     int
		expected bool
	}{
		{0, 1, true}, {0, 2, true}, {0, 3, true},
		{1, 2, true}, {1, 3, true}, {2, 3, true},
		{4, 5, true}, {4, 6, true}, {4, 7, true},
		{5, 6, true}, {5, 7, true}, {6, 7, true},
		{0, 4, false}, {1, 5, false}, {2, 6, false}, {3, 7, false},
	}

	for _, tc := range testConnections {
		if result := uf.Connected(tc.x, tc.y); result != tc.expected {
			t.Errorf("Connected(%d, %d) = %v, want %v", tc.x, tc.y, result, tc.expected)
		}
	}

	if count := uf.Count(); count != 2 {
		t.Errorf("Count() = %d, want 2", count)
	}
}

func testUnionFindSequentialOperations(t *testing.T, newUF func(n int) unionFindInterface) {
	uf := newUF(5)

	// Initial state - all separate
	if count := uf.Count(); count != 5 {
		t.Errorf("Initial Count() = %d, want 5", count)
	}

	// Check not connected
	if connected := uf.Connected(0, 1); connected {
		t.Errorf("Connected(0, 1) = %v, want false", connected)
	}

	// Union first pair
	uf.Union(0, 1)

	// Check now connected
	if connected := uf.Connected(0, 1); !connected {
		t.Errorf("Connected(0, 1) after union = %v, want true", connected)
	}

	// Count should decrease
	if count := uf.Count(); count != 4 {
		t.Errorf("Count() after first union = %d, want 4", count)
	}

	// Check Find returns same root for connected elements
	root0 := uf.Find(0)
	root1 := uf.Find(1)
	if root0 != root1 {
		t.Errorf("Find(0) = %d, Find(1) = %d, want same root", root0, root1)
	}

	// Union another pair
	uf.Union(2, 3)

	// Count should decrease again
	if count := uf.Count(); count != 3 {
		t.Errorf("Count() after second union = %d, want 3", count)
	}

	// Elements 0,1 should not be connected to 2,3
	if connected := uf.Connected(0, 2); connected {
		t.Errorf("Connected(0, 2) = %v, want false", connected)
	}
}

func testUnionFindIdempotent(t *testing.T, newUF func(n int) unionFindInterface) {
	uf := newUF(3)

	// Union the same pair multiple times
	uf.Union(0, 1)
	count1 := uf.Count()

	uf.Union(0, 1) // Should have no effect
	count2 := uf.Count()

	uf.Union(1, 0) // Reverse order, should have no effect
	count3 := uf.Count()

	if count1 != count2 || count2 != count3 {
		t.Errorf("Counts after repeated unions: %d, %d, %d - should be same", count1, count2, count3)
	}

	if count1 != 2 {
		t.Errorf("Count() = %d, want 2", count1)
	}
}

// Uncomment and implement when union-find implementation is available
func TestUnionFind(t *testing.T) {
	t.Run("Basic", func(t *testing.T) {
		testUnionFindBasic(t, func(n int) unionFindInterface { return NewUnionFind(n) })
	})
	t.Run("PathCompression", func(t *testing.T) {
		testUnionFindPathCompression(t, func(n int) unionFindInterface { return NewUnionFind(n) })
	})
	t.Run("DisjointSets", func(t *testing.T) {
		testUnionFindDisjointSets(t, func(n int) unionFindInterface { return NewUnionFind(n) })
	})
	t.Run("SequentialOperations", func(t *testing.T) {
		testUnionFindSequentialOperations(t, func(n int) unionFindInterface { return NewUnionFind(n) })
	})
	t.Run("Idempotent", func(t *testing.T) {
		testUnionFindIdempotent(t, func(n int) unionFindInterface { return NewUnionFind(n) })
	})
}
